rust/src/protocol_types/certificates/certificate.rs

```diff
@@ -0,0 +1,359 @@
+use crate::*;
+
+#[wasm_bindgen]
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub enum CertificateKind {
+    StakeRegistration,
```

Comment:
Since `RegCert` and `UnregCert` has the same semantic as `StakeRegistration` and `StakeDeregistration` we decided to combine it, because the only difference here is `coin` field. Naming such as `RegCert` and `UnregCert`  doesn't show real certificate purpose and might be confusing for a library use. When we will be sure that  `StakeRegistration` and `StakeDeregistration` will be removed in the next era, we will deprecate `StakeRegistration.new` and then delete it before hard fork.

Reply:
With decisions like these, made for backwards compatibility, you seemingly forget to account for the mental overhead of keeping this all in mind (for developers).
Developers are usually perfectly fine with updating their code, but non-obvious behaviors is what makes life a lot harder.

> Since RegCert and UnregCert has the same semantic as StakeRegistration and StakeDeregistration

You assume that the desired outcome for the developer is to *not* change the behavior while updating. But it may not be the case. IMO users should make the decisions on breaking updates, not the library, even if the changes are just API

Reply:
Going from StakeRegistration to RegCert and from StakeDeregistration to UnregCert also introduces a different workflow. Because for the lather ones you have to query the current registration deposit amount first. And for the deregistration, you have to query the used deposit amount first before generating the certificates.

Reply:
I understand that it might be confusing for developers who want to build a specific `RegCert` or `UnregCert` certificate. To address this, we have created the `new_reg_cert`, `new_unreg_cert`, `as_unreg_cert`, and `as_reg_cert` functions for the `Certificate` type.

We aim to hide this level of complexity because, generally, when a developer is building a transaction, they want to achieve a specific result or purpose (e.g., register stake credentials, spend UTXO, vote for a governance proposal). In these cases, developers do not need to concern themselves with which specific variant of the almost identical certificate or which CBOR type to use. The `StakeDeregistration` and `StakeRegistration` types reflect purpose, not just one specific certificate, even though the particular variants of how it’s serialized are changing over time, and the library might add syntactic sugar for protocol naming.

For a new developer in the Cardano ecosystem, understanding the interaction through CDDL definitions would not be the first thing, nor would it be obvious.

Introducing new types for the same operation would add to the confusion, as we cannot delete `StakeDeregistration` and `StakeRegistration` types due to the need to decode pre-Conway blocks. Adding `RegCert` and `UnregCert` as separate entities would generate more questions for developers. Even marking `StakeDeregistration` and `StakeRegistration` as deprecated might break old codebases, while the deletion of `StakeRegistration.new` would be more noticeable and easier to understand from a migration perspective.

From a stake registration perspective, the workflow will be almost the same. The only difference is where you input the deposit value—either into the transaction builder configuration or into the certificate itself—since you need to know it in any case to achieve the correct transaction balance. For deregistration, it will be almost the same, except you need to know how much deposit was paid for a particular stake registration certificate. This is valid for both the new and the old certificate variants.